src/Microsoft.DotNet.CodeFormatter.Analyzers/ExplicitVariableTypeAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+                {
+                    syntaxContext.ReportDiagnostic(Diagnostic.Create(s_ruleForEachStatement, node.Identifier.GetLocation(), node.Identifier.Text));
+                }
+            }, SyntaxKind.ForEachStatement);
+        }
+
+        /// <summary>
+        /// Returns true if given SyntaxNode has an anonymous type, which can't be replaced by an explicit type.
+        /// </summary>
+        private static bool IsAnonymousType(SyntaxNode node, SemanticModel model, CancellationToken cancellationToken)
+        {
+            ISymbol symbol = model.GetDeclaredSymbol(node, cancellationToken); 
+            bool? isAnonymousType = ((ILocalSymbol)symbol)?.Type?.IsAnonymousType;
+            return isAnonymousType.HasValue && isAnonymousType.Value;
+        }
+
+        /// <summary>
+        /// Returns true if given SyntaxNode is either VariableDeclarationSyntax or ForEachStatementSyntax and 
+        /// the variable is initialized with Non-obvious type.
+        /// 
+        ///  We define "non-obvious" as one of the following:
+        ///   1. LiteralExpressionSyntax, e.g. var x = 10;
+        ///   2. CastExpressionSyntax, e.g. var x = (Foo)f;
+        ///   3. BinaryExpressionSyntax with Kind == AsExpression
+        ///   4. A object creation syntax node, which (at least) includes:
+        ///      4.1 ObjectCreationExpressionSyntax
+        ///      4.2 ArrayCreationExpressionSyntax
+        /// 
+        ///  ImplicitArrayCreationExpressionSyntax: This one is not included. e.g. new[] {}
+        /// </summary>
+        // TODO: AnonymousObjectCreationExpressionSyntax could be filtered out here as well, maybe we want to do that?
+        //       The trade-off here is the logic we use to check syntax node is not as accurate as a query to SemanticModel, 
+        //       which is (presumbly) much slower. 
+        private static bool IsTypeObvious(SyntaxNode node)
+        {            
+            if (node == null)
+            {
+                return false;
+            }
+
+            ExpressionSyntax expressionNode = null;
+            if (node is VariableDeclarationSyntax)
+            {
+                expressionNode = ((VariableDeclarationSyntax)node)?.Variables.FirstOrDefault()?.Initializer?.Value;
+            }
+            else if (node is ForEachStatementSyntax)
+            {
+                expressionNode = ((ForEachStatementSyntax)node)?.Expression;
+            }
+            else
+            {
+                Debug.Fail("This program location is thought to be unreachable.");
+            }
+            // 'expressionNode == null' means the code under inspection has errors,
+            // so we return 'true' to avoid firing a warning.
+            return expressionNode == null ?
+                   true :
+                   expressionNode != null &&
+                 ((expressionNode is BinaryExpressionSyntax &&
+                   expressionNode.Kind() == SyntaxKind.AsExpression) ||
```

Comment:
don't you just need the SyntaxKind.AsExpression test here? IE, is it ever possible that Syntax.AsExpression is true but the item is not a BinaryExpressionSyntax? Probably not, I'm guessing